functions/src/resolve-market.ts

```diff
@@ -31,140 +40,199 @@ export const resolveMarket = functions
+  .onRun(async () => {
+    const contracts = await getValues<Contract>(
+      firestore
+        .collection('contracts')
+        .where('isResolved', '==', false)
+        .where('autoResolutionTime', '<', Date.now())
+    )
+
+    await batchedWaitAll(
+      contracts.map((contract) => async () => {
+        const result = await autoResolve(contract)
+
+        console.log('resolved', contract.slug, 'result:', result)
+      })
+    )
+  })
+
+const autoResolve = async (contract: Contract) => {
+  const data = {
+    outcome: AUTO_RESOLUTION,
+    value: undefined, // numeric
+    probabilityInt:
+      contract.outcomeType == 'BINARY'
+        ? getProbability(contract) * 100
+        : undefined,
+    resolutions:
+      contract.outcomeType == 'FREE_RESPONSE'
+        ? getFreeResponseResolutions(contract)
+        : undefined,
+  }
+  contract.description = contract.description.concat(
+    `\n\n\nContract resolved automatically.`
+  )
+
+  return await privateResolveMarket(contract, data)
+}
+
+const getFreeResponseResolutions = (contract: Contract & FreeResponse) => {
+  const answersWithProbs = getAnswersWithProbs(contract)
+  const totalProb = sum(Object.values(answersWithProbs))
+  return mapValues(answersWithProbs, (prob) => (100 * prob) / totalProb)
+}
+
+const getAnswersWithProbs = (contract: Contract & FreeResponse) => {
+  const answers: { [id: string]: number } = {}
+  for (const answer of contract.answers) {
+    answers[answer.id] = getDpmOutcomeProbability(
+      contract.totalShares,
+      answer.id
+    )
+  }
+  return answers
+}
+
+const privateResolveMarket = async (
+  contract: Contract,
+  data: {
+    outcome: resolution
+    value?: number
+    probabilityInt?: number
+    resolutions?: { [outcome: string]: number }
+  }
+) => {
+  const { creatorId, id, outcomeType, closeTime } = contract
+  const { outcome, probabilityInt, resolutions, value } = data
+  switch (outcomeType) {
+    case 'FREE_RESPONSE':
       if (
-        outcomeType === 'BINARY' &&
-        probabilityInt !== undefined &&
-        (probabilityInt < 0 ||
-          probabilityInt > 100 ||
-          !isFinite(probabilityInt))
+        isNaN(+outcome) &&
+        !(outcome === 'MKT' && resolutions) &&
+        outcome !== 'CANCEL'
       )
-        return { status: 'error', message: 'Invalid probability' }
-
-      if (creatorId !== userId)
-        return { status: 'error', message: 'User not creator of contract' }
+        return { status: 'error', message: 'Invalid outcome' }
+      break
+    case 'NUMERIC':
+      if (isNaN(+outcome) && outcome !== 'CANCEL')
+        return { status: 'error', message: 'Invalid outcome' }
+  }
+
+  if (value && !isFinite(value))
+    return { status: 'error', message: 'Invalid value' }
+
+  if (
+    outcomeType === 'BINARY' &&
+    probabilityInt &&
+    (probabilityInt < 0 || probabilityInt > 100 || !isFinite(probabilityInt))
+  )
+    return { status: 'error', message: 'Invalid probability' }
```

Comment:
This is a little suspicious. Why isn't BINARY a case in the switch statement?

I think you've lost this check from before:
```
if (outcomeType === 'BINARY') {
        if (!RESOLUTIONS.includes(outcome))
          return { status: 'error', message: 'Invalid outcome' }
}
```
I would be very careful about refactoring code like this, especially if you were not the one that wrote it initially.

Reply:
I assumed `!RESOLUTIONS.includes(outcome)` was always false.

Outcome is of type resolution:
 ```
data: {
    outcome: resolution
    value?: number
    contractId: string
    probabilityInt?: number
    resolutions?: { [outcome: string]: number }
}
```

with `resolution` and `RESOLUTIONS` being defined as.

```
export type resolution = 'YES' | 'NO' | 'MKT' | 'CANCEL'
export const RESOLUTIONS = ['YES', 'NO', 'MKT', 'CANCEL'] as const
```

Reply:
That's true — you're normally right, but this needs to be robust to rouge client calls. They might not abide by the required type, so we have to check all the parameters we have been passed to make sure they are correct.